ewatercycle/models/lisflood.py

```diff
@@ -0,0 +1,398 @@
+        2. Start bmi container and store as :py:attr:`bmi`
+
+        Args:
+            forcing: a forcing directory or a forcing data object.
+            parameterset: LISFLOOD input files. Any included forcing data will be ignored.
+            work_dir: a working directory given by user or created for user.
+
+        Returns:
+            Path to config file and path to config directory
+        """
+        singularity_image = CFG['lisflood.singularity_image']
+        docker_image = CFG['lisflood.docker_image']
+        self._check_work_dir(work_dir)
+        self._check_forcing(forcing)
+        self.parameterset = parameterset
+
+        config_file = self._create_lisflood_config()
+
+        if CFG['container_engine'].lower() == 'singularity':
+            self.bmi = BmiClientSingularity(
+                image=singularity_image,
+                input_dirs=[
+                    str(parameterset.root),
+                    str(parameterset.mask.parent),
+                    str(self.forcing_dir)
+                ],
+                work_dir=str(self.work_dir),
+            )
+        elif CFG['container_engine'].lower() == 'docker':
+            self.bmi = BmiClientDocker(
+                image=docker_image,
+                image_port=55555,
+                input_dirs=[
+                    str(parameterset.root),
+                    str(parameterset.mask.parent),
+                    str(self.forcing_dir)
+                ],
+                work_dir=str(self.work_dir),
+            )
+        else:
+            raise ValueError(
+                f"Unknown container technology in CFG: {CFG['container_engine']}"
+            )
+        return Path(config_file), self.work_dir
+
+    def _check_work_dir(self, work_dir):
+        """Check and optionally create the work directory."""
+        self.work_dir = _generate_workdir(work_dir)
+
+    def _check_forcing(self, forcing):
+        """"Check forcing argument."""
+
+        # TODO check if mask has same grid as forcing files,
+        # if not warn users to run reindex_forcings
+        if isinstance(forcing, PathLike):
+            self.forcing_dir = forcing.expanduser().resolve()
+            self.forcing_files = dict()
+            for forcing_file in self.forcing_dir.glob('*.nc'):
+                dataset = xr.open_dataset(forcing_file)
+                # TODO check dataset was created by ESMValTool, to make sure var names are as expected
+                var_name = list(dataset.data_vars.keys())[0]
+                self.forcing_files[var_name] = forcing_file.name
+                # get start and end date of time dimension
+                # TODO converting numpy.datetime64 to datetime object is ugly, find better way
+                self.start = datetime.utcfromtimestamp(dataset.coords['time'][0].values.astype('O') / 1e9)
+                self.end = datetime.utcfromtimestamp(dataset.coords['time'][-1].values.astype('O') / 1e9)
```

Comment:
Maybe something to think about: do we want the forcing data to determine the start/end time, or do we want the used to set the start/end time independently of the forcing?

Reply:
Good question. If the forcing is the only temporal data given to lisflood, it makes sense that the start/end time of forcing is used in configuration and thus in initialization. Users can choose any start/end time (in forcing time range) while runs the model.

Reply:
The forcing gives a maximum range. A user can use update() to stop early, but now can not start later. Supplying a start date to setup could help, but not in scope for current use cases.

Reply:
We added a TODO item in setup().